src/NuGet.Core/NuGet.Packaging/Signing/Authoring/X509SignatureProvider.cs

```diff
@@ -71,9 +76,27 @@ private Task<Signature> CreateSignature(X509Certificate2 cert, SignatureManifest
                     new AsnEncodedDataCollection(signingTime)));
 
             cmsSigner.IncludeOption = X509IncludeOption.WholeChain;
+            cmsSigner.DigestAlgorithm = cert.SignatureAlgorithm;
 
             var cms = new SignedCms(contentInfo);
-            cms.ComputeSignature(cmsSigner);
+
+            try
+            {
+                cms.ComputeSignature(cmsSigner);
+            }
+            catch (CryptographicException ex)
```

Comment:
:sunflower: a `when` here would remove the need for the switch case and default throw

`catch (CryptographicException ex) when (ex.HResult == INVALID_PROVIDER_TYPE_HRESULT)`